Add ESQL query performance guide (#150360)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-query-performance.md

| [`CASE`](/reference/query-languages/esql/functions-operators/conditional-functions-and-expressions/case.md) | Conditional aggregation through `CASE` | Lazy evaluation, slow |
| [`LOOKUP JOIN`](/reference/query-languages/esql/commands/lookup-join.md) | Join against a large lookup index | Cost is proportional to the lookup index size |

:::{tip}
The most impactful fixes are usually: add a time range filter, add a [`WHERE`](/reference/query-languages/esql/commands/where.md), and add a [`KEEP`](/reference/query-languages/esql/commands/keep.md).
:::

## Reduce what you scan

Most {{esql}} queries spend the bulk of their time reading data from disk. The fastest queries read the least data. This section covers the four levers that most directly control scan size.

### Narrow the time range

A tight time range is the single biggest performance lever in most workloads. {{esql}} uses the `@timestamp` field to skip entire shards and segments that fall outside the range, so a narrower window directly reduces the amount of data read.

Avoid running queries that span more time than the result actually needs:

> ❌ **Don't**: Query without a time bound

Reviewer: ```suggestion
❌ **Don't:** Query without a time bound
```

Nit: Can you include the colon in the same formatting as Do and Don't?

```

Use wildcards in [`KEEP`](/reference/query-languages/esql/commands/keep.md) sparingly. `host.*` is better than no `KEEP` at all, but `host.name` is better than `host.*` because it avoids pulling in adjacent fields.

### Cap rows with LIMIT

Always include a [`LIMIT`](/reference/query-languages/esql/commands/limit.md) on queries that return raw rows. Large result sets cause slow serialization and can trigger deserialization errors in {{kib}}.

❌ **Don't**: Leave the result set unbounded
```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day AND log.level == "error"
| SORT @timestamp DESC
```

✅ **Do**: Cap the result to the rows the consumer actually needs

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day AND log.level == "error"
| SORT @timestamp DESC
| LIMIT 100
```

## Avoid expensive operations

Some {{esql}} operations are intrinsically more expensive than their alternatives. Knowing the cheaper substitute, and when it applies, often replaces a slow query with a fast one. The subsections below are ordered roughly by impact, with the highest-leverage changes first.

### Use full-text search instead of LIKE or RLIKE

```{applies_to}
stack: ga 8.17
serverless: ga

> ```

Author: ```suggestion
```

> For text search, prefer [`MATCH`](/reference/query-languages/esql/functions-operators/search-functions/match.md), [`QSTR`](/reference/query-languages/esql/functions-operators/search-functions/qstr.md), or [`KQL`](/reference/query-languages/esql/functions-operators/search-functions/kql.md) over [`LIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-like) or [`RLIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-rlike). The full-text search functions use the inverted index and are the right fit for analyzed text. `LIKE` and `RLIKE` are pattern-matching operators; pre 8.18/9.0 they are especially costly because they are not pushed down to Lucene.

Reviewer: also mention match_phrase

❌ **Don't**: Use pattern matching on free text with [`LIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-like)
```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour
  AND message LIKE "*connection refused*"
```

✅ **Do**: Use [`MATCH`](/reference/query-languages/esql/functions-operators/search-functions/match.md) against the inverted index

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour

>   AND MATCH(message, "connection refused") // Inverted index lookup, not a row scan

Reviewer: This should probably be match_phrase so that it matches "connection refused" as a pair

```

[`MATCH`](/reference/query-languages/esql/functions-operators/search-functions/match.md) works on `text` and `keyword` fields. For Lucene query syntax with `field:value` and boolean operators, use [`QSTR`](/reference/query-languages/esql/functions-operators/search-functions/qstr.md). For {{kib}} Query Language syntax, use [`KQL`](/reference/query-languages/esql/functions-operators/search-functions/kql.md).

:::{tip}

> To learn more about full-text search patterns in {{esql}}, refer to [{{esql}} for search](docs-content://solutions/search/esql-for-search.md).

Author: ```suggestion
To learn more about using {{esql}} for search use cases, refer to [{{esql}} for search](docs-content://solutions/search/esql-for-search.md).
```

```esql
FROM metrics-system.cpu-*
| WHERE @timestamp > NOW() - 1 hour
| STATS avg_cpu = AVG(system.cpu.user.pct) BY host.name, bucket = DATE_TRUNC(5 minutes, @timestamp)
```

✅ **Do:** Use [`TS`](/reference/query-languages/esql/commands/ts.md) with [`TBUCKET`](/reference/query-languages/esql/functions-operators/grouping-functions/tbucket.md) for time series metrics

```esql
TS metrics-system.cpu-*
| STATS avg_cpu = AVG(AVG_OVER_TIME(system.cpu.user.pct)) // Inner-then-outer aggregation pattern
        BY host.name, TBUCKET(5 minutes)                  // TBUCKET replaces DATE_TRUNC under TS
```

:::{important}
[`TS`](/reference/query-languages/esql/commands/ts.md) only works on indices created as time series data streams. For non-TSDS indices, continue to use [`FROM`](/reference/query-languages/esql/commands/from.md).
:::

## Reduce what you return

Every column returned has to be read from storage, serialized, and transmitted. Shrinking the result set, by returning fewer columns or rows, often produces significant gains on large indices.

### Select columns with KEEP

> [`KEEP`](/reference/query-languages/esql/commands/keep.md) selects which columns to return. [`DROP`](/reference/query-languages/esql/commands/drop.md) does the inverse. Without either, {{esql}} returns every field in every matching document. This is the single biggest source of avoidable overhead on indices with hundreds or thousands of fields.

Reviewer 2: We also have a "drop_null_columns" query parameter, that can make a significant difference on the result serialisation side, especially on very sparse datasets.

❌ **Don't:** Return every field by default
```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour AND log.level == "error"
| SORT @timestamp DESC
| LIMIT 100
```

✅ **Do:** Project only the fields the consumer actually needs

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour AND log.level == "error"
| KEEP @timestamp, host.name, message, log.level // Return only the four fields needed downstream
| SORT @timestamp DESC
| LIMIT 100
```

:::{tip}
Use wildcards in [`KEEP`](/reference/query-languages/esql/commands/keep.md) sparingly. `host.*` is better than no `KEEP` at all, but `host.name` is better than `host.*` because it avoids pulling in adjacent fields.
:::

### Cap rows with LIMIT

Always include a [`LIMIT`](/reference/query-languages/esql/commands/limit.md) on queries that return raw rows. Large result sets cause slow serialization and can trigger deserialization errors in {{kib}}.

> ❌ **Don't:** Leave the result set unbounded

Reviewer 2: Worth mentioning that ES|QL adds a default limit of 1000 at the end of every query. 
The best thing you can do is to reduce it with an explicit LIMIT.
The worst thing you can do it increase it with a broader LIMIT (but ES|QL still protects you with a default max-limit of 10k).

| SORT @timestamp DESC
| LIMIT 100
```

## Avoid expensive operations

Some {{esql}} operations are intrinsically more expensive than their alternatives. Knowing the cheaper substitute, and when it applies, often replaces a slow query with a fast one. The subsections below are ordered roughly by impact, with the highest-leverage changes first.

### Use full-text search instead of LIKE or RLIKE

For text search, prefer [`MATCH`](/reference/query-languages/esql/functions-operators/search-functions/match.md), [`MATCH_PHRASE`](/reference/query-languages/esql/functions-operators/search-functions/match_phrase.md), [`QSTR`](/reference/query-languages/esql/functions-operators/search-functions/qstr.md), or [`KQL`](/reference/query-languages/esql/functions-operators/search-functions/kql.md) over [`LIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-like) or [`RLIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-rlike). The full-text search functions use the inverted index and are optimized for analyzed text. `LIKE` and `RLIKE` are pattern-matching operators. Pre 8.18/9.0 they are especially costly because they are not pushed down to Lucene.

> ❌ **Don't:** Use pattern matching on free text with [`LIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-like)

Reviewer 2: Worth mentioning that leading wildcards (eg. `*something`) are particularly expensive.

### Prefer fields backed by doc values

{{esql}} reads field values through a block-loading system that strongly prefers [doc values](/reference/elasticsearch/mapping-reference/doc-values.md). Fields with doc values, such as `keyword`, [numeric](/reference/elasticsearch/mapping-reference/number.md), `date`, and `ip` types, are read in fast columnar batches. Fields without doc values, such as `text` and `match_only_text`, fall back to reading [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md), which requires decompressing and parsing the full JSON document per row. This applies to any operation that reads the field value, including filtering, grouping, sorting, and returning fields through [`KEEP`](/reference/query-languages/esql/commands/keep.md).

When a `.keyword` subfield exists, prefer it for filtering, grouping, and returning. When one does not, filter aggressively to limit the number of documents that require `_source` reads.

❌ **Don't:** Group by an analyzed field
```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour
| STATS count = COUNT(*) BY message
```

> ✅ **Do:** Use the `.keyword` subfield

Reviewer 2: It's more subtle than this: if an _exact_ keyword subfield exists, the planner will rewrite the expressions for you, so the two will have the same performance.
If the text field has no keyword subfield, or if it's not exact (eg. if it has `ignore_above`), then the planner will ignore it, and it will fetch the text from source, and _that_ will be definitely expensive.

---
applies_to:
  stack: ga
  serverless: ga
products:
  - id: elasticsearch
navigation_title: "Optimize query performance"

> description: "Techniques for writing fast ES|QL queries and operating ES|QL workloads at scale."

Reviewer 3: we could add `type: overview` to the frontmatter

Author: it's somewhere between a how-to and tutorial perhaps

---

# Optimize {{esql}} query performance

> This guide covers practical techniques for writing fast {{esql}} queries and operating {{esql}} workloads at scale. The first sections focus on writing efficient queries. The later sections cover monitoring and operational tools that help identify slow queries in production.

Reviewer 3: > The first sections focus on writing efficient queries. The later sections cover monitoring and operational tools that help identify slow queries in production.

At first glance, it's a little difficult to tell which sections focus on efficient queries and which ones cover operational tools. Perhaps we could either replace this sentence with a more general introduction, or make it a bit more specific by briefly describing the sections covered on the page.

Author: good call 👍

::::{tip}
For a quick overview of the most common issues with {{esql}} queries, refer to [Common anti-patterns](#common-anti-patterns).
::::

## Before you begin

This guide assumes familiarity with {{esql}} syntax and command pipelines. To learn the basics, refer to [Get started with {{esql}} queries](/reference/query-languages/esql/esql-getting-started.md).

This guide serves two audiences:

- If you write {{esql}} queries, the following sections cover the techniques most relevant to you:
  - [Common anti-patterns](#common-anti-patterns)
  - [Reduce what you scan](#reduce-what-you-scan)
  - [Reduce what you return](#reduce-what-you-return)
  - [Avoid expensive operations](#avoid-expensive-operations)

> - If you administer clusters and need to monitor query performance fleet-wide, the following section describes the tools available:

Reviewer 3: Instead of "fleet-wide", I wonder if something more explicit, such as "across multiple clusters", might be easier for non-native readers to understand

Author: 💯

  - [Monitor query performance](#monitor-query-performance)

### Check your Elastic Stack version

If you're not on [{{serverless-full}}](docs-content://deploy-manage/deploy/elastic-cloud/serverless.md), check your {{stack}} version. The {{esql}} query engine improves with each release, so upgrading is often one of the highest-impact performance changes you can make.

Some tips on this page require a recent version of the {{stack}}, and individual subsections carry an applicability badge when this is the case. Sections without a version badge apply to all versions where {{esql}} is generally available.

> For clusters on a version before 8.17, upgrading provides the largest single performance improvement, because full-text search functions and Lucene pushdowns become available. For clusters on 8.17 but before 8.18, upgrading to 8.18 provides the next largest improvement. That release adds [`LIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-like) and [`RLIKE`](/reference/query-languages/esql/functions-operators/operators.md#esql-rlike) pushdown to Lucene and a mapping discovery optimization that reduces overhead on clusters with many indices.

Reviewer 4: This section seems to focus only on the 8.x series. At first I thought there was no 9.x info, until I expanded the version specific section below, which is collapsed by default. That seems to make the 9.x stuff appear less important, or invisible if you don't know to expand the box.

## Common anti-patterns

These anti-patterns are the most common causes of {{esql}} query latency in production.

:::{tip}
:applies_to: { ech:, serverless: }

[AutoOps](docs-content://deploy-manage/monitor/autoops.md) detects most of these patterns automatically and surfaces actionable recommendations. To browse detected events, refer to [AutoOps events](docs-content://deploy-manage/monitor/autoops/ec-autoops-events.md).
:::

| Pattern | What to look for | Why it's slow |
| :---- | :---- | :---- |
| Broad index pattern | [`FROM *`](/reference/query-languages/esql/commands/from.md) or wide wildcards | Expensive mapping discovery, plus scans across many indices |
| Wide time range | `@timestamp` range spanning weeks or months | Scans proportionally more data |
| Missing [`WHERE`](/reference/query-languages/esql/commands/where.md) | No filter conditions at all | Full index scan |
| Missing [`KEEP`](/reference/query-languages/esql/commands/keep.md) | No column selection | Returns all fields, producing large payloads |
| Missing [`LIMIT`](/reference/query-languages/esql/commands/limit.md) | Unbounded result set | Slow serialization, can trigger deserialization errors in {{kib}} |

> | High-cardinality [`STATS`](/reference/query-languages/esql/commands/stats-by.md) `BY` | Grouping by raw timestamps, full URLs, or document IDs | Produces millions of buckets, can trip circuit breakers |

Reviewer 4: I think the `BY` should be inside the link, not outside.

Most {{esql}} queries spend the bulk of their time reading data from disk. The fastest queries read the least data. This section covers the levers that most directly control scan size.

### Narrow the time range

A tight time range is the single biggest performance lever in most workloads. {{esql}} uses the `@timestamp` field to skip entire shards and segments that fall outside the range, so a narrower window directly reduces the amount of data read.

Avoid running queries that span more time than the result actually needs:

❌ **Don't:** Query without a time bound
```esql
FROM logs-*
| WHERE host.name == "web-01"
| STATS count = COUNT(*) BY log.level
```

✅ **Do:** Add an explicit `@timestamp` filter to bound the scan

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day // Bound the scan to a tight window
  AND host.name == "web-01"
| STATS count = COUNT(*) BY log.level
```

In {{kib}}, the time picker automatically applies a time range filter. When writing queries directly in the [{{kib}} Console](docs-content://explore-analyze/query-filter/tools/console.md) or through the API, always include an explicit `@timestamp` filter.

### Filter early with WHERE

> A [`WHERE`](/reference/query-languages/esql/commands/where.md) clause earlier in the pipeline reduces the dataset before downstream commands process it. Conditions on indexed fields such as `keyword`, [numeric](/reference/elasticsearch/mapping-reference/number.md), `date`, or `ip` types are pushed down to Lucene, which skips irrelevant documents entirely.

Reviewer 4: Should we be more inclusive in the types that can be filtered by Lucene? For example, lucene pushdown is certainly support for all the spatial field types: geo_point, geo_shape, cartesian_point and cartesian_shape.

Without a [`WHERE`](/reference/query-languages/esql/commands/where.md), {{esql}} scans every document in the matched indices:

> ❌ **Don't:** Filter after the aggregation

Reviewer 5: TIL: https://github.com/elastic/elasticsearch/issues/115311
Automatic pushdown of filter past stats is still not fully implemented :(

Reviewer 4: Really! That's a shocker!

```esql
FROM logs-*
| STATS count = COUNT(*) BY host.name, log.level
| WHERE log.level == "error"
```

✅ **Do:** Push the filter up so it runs before the aggregation

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day // Filter pushed to Lucene
  AND log.level == "error"
| STATS count = COUNT(*) BY host.name
```

### Restrict the index pattern

> A broad [`FROM *`](/reference/query-languages/esql/commands/from.md) forces {{esql}} to discover field mappings across every index and then query each one. On clusters with thousands of indices, that discovery overhead alone can dominate query time.

Reviewer 5: This is especially expensive with CPS where `*` expands to all linked projects.
I wonder if it is worth highlighting that case? Possibly referencing project routing to restricting index resolution to origin only

Author: great point, i'll add that 👍

:::{warning}
High-cardinality groupings can exhaust memory and trip circuit breakers. Always bucket timestamps and choose the lowest-cardinality representation of a field that still answers the question.
:::

❌ **Don't:** Group by raw, high-cardinality fields
```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day
| STATS count = COUNT(*) BY url.full, user.name, @timestamp
```

✅ **Do:** Reduce cardinality before grouping

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 day
| STATS count = COUNT(*) BY url.path, user.name, bucket = DATE_TRUNC(1 hour, @timestamp) // Uses `url.path` instead of `url.full`, bucketed timestamps instead of raw
```

Common reductions include: bucketing timestamps with [`DATE_TRUNC`](/reference/query-languages/esql/functions-operators/date-time-functions/date_trunc.md) or [`BUCKET`](/reference/query-languages/esql/functions-operators/grouping-functions/bucket.md), using `url.path` instead of `url.full`, and filtering to a known subset before the [`STATS`](/reference/query-languages/esql/commands/stats-by.md).

### Prefer fields backed by doc values

> {{esql}} reads field values through a block-loading system that strongly prefers [doc values](/reference/elasticsearch/mapping-reference/doc-values.md). Fields with doc values, such as `keyword`, [numeric](/reference/elasticsearch/mapping-reference/number.md), `date`, and `ip` types, are read in fast columnar batches. Fields without doc values, such as `text` and `match_only_text`, fall back to reading [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md), which requires decompressing and parsing the full JSON document per row. This applies to any operation that reads the field value, including filtering, grouping, sorting, and returning fields through [`KEEP`](/reference/query-languages/esql/commands/keep.md).

Reviewer 5: This contradicts above

> If you never sort or aggregate on a field, disable [`doc_values`](/reference/elasticsearch/mapping-reference/doc-values.md) instead. Every structure you remove is one less thing {{esql}} has to read.

I suggest to remove quoted paragraph as that is mostly indexing overhead.
As suggested above, doc values could speed up reading when values are needed. When not, extra columns could be removed using keep/drop.

Reviewer 4: I also think we should have a section on spatial types and doc-values. Unlike other types, the spatial types loose information in the doc values, and we're forced to read from source (very slow) if the user actually returns the original field in the results. There are ways around this, and they are worth documenting. Perhaps a followup PR to add to this content?

Reviewer 4: For non-spatial fields, ES|QL prefers to use doc-values all the time, so I agree with Ievgens suggestion to remove the paragraph recommending to not store doc-values.

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour
| GROK message "%{TIMESTAMP_ISO8601:ts} %{LOGLEVEL:level} %{GREEDYDATA:msg}"
```

✅ **Do:** Use [`DISSECT`](/reference/query-languages/esql/commands/dissect.md) for delimiter-based formats

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour
| DISSECT message "%{ts} %{level} %{msg}" // Delimiter tokenization, no regex engine
```

### Filter before LOOKUP JOIN

[`LOOKUP JOIN`](/reference/query-languages/esql/commands/lookup-join.md) combines each incoming row with matching rows from a lookup index. Joining fewer incoming rows is usually faster, and large lookup matches can increase memory pressure.

Filter the source data before joining, and keep the lookup index as small and purpose-built as possible:

```esql
FROM logs-*
| WHERE @timestamp > NOW() - 1 hour
  AND event.category == "network"
| LOOKUP JOIN threat_list ON source.ip
| KEEP @timestamp, source.ip, threat_list.risk, event.action
```

{{esql}} tries to push filters before the join when possible. Write the query with the selective filters before `LOOKUP JOIN` so the intended execution order is clear and the join receives the smallest practical input.

> ### Use approximate aggregations where possible

Reviewer 4: Should we use an applies_to pill on this section?
